src/cuda-poh-verify/poh_verify.cu

```diff
@@ -0,0 +1,151 @@
+#include <stddef.h>
+#include <inttypes.h>
+#include <pthread.h>
+#include "gpu_common.h"
+#include "sha256.h"
+
+#define MAX_NUM_GPUS 8
+#define MAX_QUEUE_SIZE 8
+#define NUM_THREADS_PER_BLOCK 64
+
+
+__global__ void poh_verify_kernel(uint8_t* hashes, uint64_t* num_hashes_arr, size_t num_elems) {
+    size_t idx = (size_t)(blockIdx.x * blockDim.x + threadIdx.x);
+    if (idx >= num_elems) return;
+
+    uint8_t hash[SHA256_BLOCK_SIZE];
+
+    memcpy(hash, &hashes[idx * SHA256_BLOCK_SIZE], SHA256_BLOCK_SIZE);
+
+    for (size_t i = 0; i < num_hashes_arr[idx]; i++) {
+        hash_state sha_state;
+        sha256_init(&sha_state);
+        sha256_process(&sha_state, hash, SHA256_BLOCK_SIZE);
+        sha256_done(&sha_state, hash);
+    }
+
+    memcpy(&hashes[idx * SHA256_BLOCK_SIZE], hash, SHA256_BLOCK_SIZE);
+}
+
+typedef struct {
+    uint8_t* hashes;
+    uint64_t* num_hashes_arr;
+    size_t num_elems_alloc;
+    pthread_mutex_t mutex;
+    cudaStream_t stream;
+} gpu_ctx;
+
+static pthread_mutex_t g_ctx_mutex = PTHREAD_MUTEX_INITIALIZER;
+
+static gpu_ctx g_gpu_ctx[MAX_NUM_GPUS][MAX_QUEUE_SIZE] = {0};
+static uint32_t g_cur_gpu = 0;
+static uint32_t g_cur_queue[MAX_NUM_GPUS] = {0};
+static int32_t g_total_gpus = -1;
+
+static bool poh_init_locked() {
+    if (g_total_gpus == -1) {
+        cudaGetDeviceCount(&g_total_gpus);
+        g_total_gpus = min(MAX_NUM_GPUS, g_total_gpus);
+        LOG("total_gpus: %d\n", g_total_gpus);
+        for (int gpu = 0; gpu < g_total_gpus; gpu++) {
+            CUDA_CHK(cudaSetDevice(gpu));
+            for (int queue = 0; queue < MAX_QUEUE_SIZE; queue++) {
+                int err = pthread_mutex_init(&g_gpu_ctx[gpu][queue].mutex, NULL);
+                if (err != 0) {
+                    fprintf(stderr, "pthread_mutex_init error %d gpu: %d queue: %d\n",
+                            err, gpu, queue);
+                    g_total_gpus = 0;
+                    return false;
+                }
+                CUDA_CHK(cudaStreamCreate(&g_gpu_ctx[gpu][queue].stream));
+            }
+        }
+    }
+    return g_total_gpus > 0;
+}
+
+bool poh_init() {
+    cudaFree(0);
+    pthread_mutex_lock(&g_ctx_mutex);
+    bool success = poh_init_locked();
+    pthread_mutex_unlock(&g_ctx_mutex);
+    return success;
+}
+
+// This function locks the pthreads mutex for the gpu context it takes
+// for the computation. `poh_finish_verify_many` must be called with the
+// returned gpu_ctx to avoid poisoning the mutex. This structure allows
+// the host to complete calculations concurrently with the gpu calculations.
```

Comment:
You may want to do all this stuff in another cpu thread anyway to parallelize the driver work, but this works too. Not 100% sure which would be faster.

Reply:
See first reply.